Utility computation when: custom group sizes and columns in a group.

mabhijithn opened the pull request.

mabhijithn committed:
Added input argument groupid and error check

mabhijithn committed:
Function to calculate group utility. Custom group sizes and arrangements possible

servaasvdc commented on channel_select.m:
```diff
@@ -55,6 +55,12 @@
+                    nofuniq = unique(grpid); % Number of unique groups
+                    if(size(grpid,1)~=size(A,2))
+                        error('Length of group-ids != No.of columns in A');
+                    end
```
Consider using [inputParser](https://nl.mathworks.com/help/matlab/ref/inputparser.html) for all this argument processing stuff

servaasvdc commented on grputilcalc.m:
```diff
@@ -0,0 +1,31 @@
+%          columns of data with each entry containing the group ID that the
+%          corresponding column in A belongs to.
+function util = grputilcalc (RXXinv, W, grpid)
+
```
Have a look at https://nl.mathworks.com/help/matlab/matlab_prog/add-help-for-your-program.html for info on how to document a function. `help grputilcalc` will look probably look whack like this

servaasvdc replied:
Apart from the formatting, some examples would be great as well. See also https://nl.mathworks.com/matlabcentral/fileexchange/46056-matlab-style-guidelines-2-0 -- great guide.

servaasvdc commented on grputilcalc.m:
```diff
@@ -0,0 +1,31 @@
+function util = grputilcalc (RXXinv, W, grpid)
+
+    uniqgrps = unique(grpid);
+    util = zeros(length(uniqgrps),1);
```
`util = zeros(size(uniqgrps))` (unless you really need it to be a column vector)

mabhijithn replied:
Yes. I need it to be a column vector in later part of channel_select.m

servaasvdc commented on grputilcalc.m:
```diff
@@ -0,0 +1,31 @@
+
+    uniqgrps = unique(grpid);
+    util = zeros(length(uniqgrps),1);
+    for k = 1:length(uniqgrps)
```
You could do `for uniqgrpId = uniqgrps` directly instead of indexing later

mabhijithn replied:
The util vector needs updating from 1 to length(uniqgrps), so an indexing variable is required anyway.

servaasvdc commented on grputilcalc.m:
```diff
@@ -0,0 +1,31 @@
+    uniqgrps = unique(grpid);
+    util = zeros(length(uniqgrps),1);
+    for k = 1:length(uniqgrps)
+        indx = (grpid==uniqgrps(k));
+        
+        % Select decoder weights of group k 
+        Wkq = W(indx);
+        
+        % utility computation of group k
+        S = RXXinv(indx,indx);
+        
+        util(k) = Wkq'*(S\Wkq);
```
Looks okay but you could probably do this in a single line with [splitapply](https://nl.mathworks.com/help/matlab/ref/splitapply.html), where `grpid` is the grouping variable

mabhijithn committed:
Modified help to be more in-line with convention.

servaasvdc reviewed (commented):
Looks okay for now. Lemme know if there's anything else you want me to take a look at

mabhijithn committed:
Support for custom group-IDs started. Not fully supported but normal channel-selection now in group-ID format

Commit:
Merge branch 'customgrps' of github.com:mabhijithn/channelselect into customgrps

Commit:
Replaced inv with backslash

mabhijithn committed:
Loop fixed to select N<total-channels and N=tota-channels

mabhijithn committed:
- Changed check of singualrity with condition number check; faster check
- Changed lambda selection to minimum positive eigen value; to match with the theory in paper

mabhijithn committed:
Recursive computation of inverse added

mabhijithn committed:
Bug fix: normal inverse when not redundant

mabhijithn committed:
Bug fix: removed try catch block
